py-polars/polars/internals/construction.py

```diff
@@ -32,33 +32,30 @@
 )
 from polars.utils import threadpool_size
 
+if TYPE_CHECKING:  # pragma: no cover
+    import pandas as pd
+
```

Comment:
Pyarrow shold be explicitly imported during typechecking.

Reply:
Why is that? I am assuming our development dependencies (which include pyarrow) are available whenever we are doing type checking. So the import as it is now should be sufficient.

If we do need to import the optional dependencies explicitly during type checking, we would have to do this for numpy and pandas as well.

Reply:
It has to be done for pandas and numpy as well. Just in case you have an environment with only polars, and run the tests (if you fixed a bug).

Reply:
To me, it feels like we are introducing unnecessary complexity in the imports that way. I feel it is reasonable to expect that your development dependencies are available during testing. And in the case you describe, the only difference you will find is that your type checking will fail for a different reason.

Let's wait and see what Ritchie thinks. Easy enough to incorporate that logic in a new commit if we do decide to go for that.

Reply:
Maybe the `if TYPE_CHECKING` parts are not needed anymore after https://github.com/pola-rs/polars/pull/3863 is merged.

Reply:
I will rebase this after that PR merged, maybe you could test out your case to see if it works as expected?

Reply:
The type annotation PR has been merged now - could you check if this branch now behaves correctly in the case you mentioned?

Reply:
I see that this branch has now been merged. Should any issues arise when type hinting, I would propose adding the following clause below the optional imports:

```python
if TYPE_CHECKING and not (_NUMPY_AVAILABLE and _PYARROW_AVAILABLE):
    raise ImportError("Optional dependencies must be available during type checking.")
```